pytradfri/api/aiocoap_api.py

```diff
@@ -0,0 +1,162 @@
+"""Coap implementation using aiocoap."""
+import asyncio
+import json
+import logging
+import sys
+
+sys.path.insert(0, '/usr/src/build/tinydtls/cython')  # noqa
+
+import aiocoap
+from aiocoap import Message, Context
+from aiocoap.error import RequestTimedOut
+from aiocoap.numbers.codes import Code
+from aiocoap.transports.tinydtls import TransportEndpointTinyDTLS
+
+from ..error import ClientError, ServerError, RequestTimeout
+from ..command import Command
+
+_LOGGER = logging.getLogger(__name__)
+
+aiocoap.numbers.constants.MAX_RETRANSMIT = 10
+
+
+class PatchedDTLSSecurityStore:
+    """Patched DTLS store in lieu of impl."""
+
+    SECRET_PSK = None
+
+    def _get_psk(self, host, port):
+        return b"Client_identity", PatchedDTLSSecurityStore.SECRET_PSK
+
+
+aiocoap.transports.tinydtls.DTLSSecurityStore = PatchedDTLSSecurityStore
+
+
+def _patched_datagram_received(self, data, addr):
+    self._dtls_socket.handleMessage(self._connection, data, 0)
+
+
+aiocoap.transports.tinydtls.DTLSClientConnection.datagram_received = \
+    _patched_datagram_received
+
+
+@asyncio.coroutine
+def api_factory(host, security_code):
+    """Generate a request method."""
+
+    security_code = security_code.encode('utf-8')
+
+    PatchedDTLSSecurityStore.SECRET_PSK = security_code
+
+    @asyncio.coroutine
+    def _get_protocol():
+        protocol = yield from Context.create_client_context()
+        for endpoint in protocol.transport_endpoints:
+            if type(endpoint) == TransportEndpointTinyDTLS:
+                pass
+        return protocol
+
+    @asyncio.coroutine
+    def _execute(api_command):
+        """Execute the command."""
+        if api_command.observe:
+            yield from _observe(api_command)
+            return
+
+        method = api_command.method
+        path = api_command.path
+        data = api_command.data
+        parse_json = api_command.parse_json
+        url = api_command.url(host)
+
+        kwargs = {}
+
+        if data is not None:
+            kwargs['payload'] = json.dumps(data).encode('utf-8')
+            _LOGGER.debug('Executing %s %s %s: %s', host, method, path, data)
+        else:
+            _LOGGER.debug('Executing %s %s %s', host, method, path)
+
+        api_method = Code.GET
+        if method == 'put':
+            api_method = Code.PUT
+
+        msg = Message(code=api_method, uri=url, **kwargs)
+
+        try:
+            protocol = yield from _get_protocol()
+            res = yield from protocol.request(msg).response
+        except RequestTimedOut:
+            raise RequestTimeout('Request timed out.')
+
+        api_command.result = _process_output(res, parse_json)
+
+        return api_command.result
+
+    @asyncio.coroutine
+    def request(*api_commands):
+        """Make a request."""
+        if len(api_commands) == 1:
+            result = yield from _execute(api_commands[0])
+            return result
+
+        command_results = []
+
+        for api_command in api_commands:
```

Comment:
Is there any reason why we would do this in sequence instead of in parallel?

```python
command_results = yield from asyncio.gather(
    *(_execute(api_command) for api_command in api_commands), loop=loop)
```

(this will require to pass `loop` into api_factory and default to `loop = asyncio.get_event_loop()`)